### PR TITLE
feat: add --sysctl-init-file as cli command option

### PR DESCRIPTION
**Pull request by alexruzenhack**
### Acceptance criteria
- Add `--sysctl-init-file` CLI command option that should use the sysctl subsystem to configure the node.
- The configuration file is a txt file with a simple key value pair scheme.

Closes:
- https://github.com/HathorNetwork/hathor-core/issues/684
- https://github.com/HathorNetwork/hathor-core/issues/685

### CODE REVIEW

File: hathor/builder/sysctl_builder.py [hunk 1]
```diff
@@ -30,4 +41,26 @@ def build(self) -> Sysctl:
         if ws_factory is not None:
             root.put_child('ws', WebsocketManagerSysctl(ws_factory))
 
+        if self.sysctl_config_file:
+            # load and parse init file
```
Comment (alexruzenhack) [inline hathor/builder/sysctl_builder.py:45]: remove

File: hathor/builder/sysctl_builder.py [hunk 2]
```diff
@@ -30,4 +41,26 @@ def build(self) -> Sysctl:
...
+            sysctl_settings_dict = self._load_yaml(self.sysctl_config_file)
+            allowed_commands = set(root.get_all_paths())
+            settings = set(sysctl_settings_dict.keys())
+            self._validate_sysctl_settings(settings, allowed_commands)
+
+            # configure sysctl
```
Comment (alexruzenhack) [inline hathor/builder/sysctl_builder.py:51]: Suggesting: # apply settings using sysctl

File: hathor/builder/sysctl_builder.py [hunk 3]
```diff
@@ -30,4 +41,26 @@ def build(self) -> Sysctl:
...
+            for command, value in sysctl_settings_dict.items():
+                root.set(command, value)
+
         return root
+
+    def _load_yaml(self, path: str) -> dict:
```
Comment (alexruzenhack) [inline hathor/builder/sysctl_builder.py:57]: Add docstring

File: hathor/builder/sysctl_builder.py [hunk 4]
```diff
@@ -30,4 +41,26 @@ def build(self) -> Sysctl:
...
+        root_dir = Path(__file__).parent.parent
+        return yaml.dict_from_yaml(
+            filepath=root_dir.joinpath(path),
+        )
+
+    def _validate_sysctl_settings(self, settings: set[str], commands: set[str]) -> None:
```
Comment (alexruzenhack) [inline hathor/builder/sysctl_builder.py:63]: Add docstring

File: hathor/cli/run_node.py [hunk 5]
```diff
@@ -58,6 +58,8 @@ def create_parser(cls) -> ArgumentParser:
         parser.add_argument('--peer', help='json file with peer info')
         parser.add_argument('--sysctl',
                             help='Endpoint description (eg: unix:/path/sysctl.sock, tcp:5000:interface:127.0.0.1)')
+        parser.add_argument('--sysctl-config-file',
+                            help='File path to the sysctl.yml config file (eg: conf/sysctl.)')
```
Comment (alexruzenhack) [inline hathor/cli/run_node.py:62]: fix: "conf/sysctl.yml"

File: hathor/cli/run_node.py [hunk 6]
```diff
@@ -357,18 +359,22 @@ def __init__(self, *, argv=None):
...
+            if self._args.sysctl_config_file:
+                self.init_sysctl(
+                    self._args.sysctl, self._args.sysctl_config_file)
+            else:
+                self.init_sysctl(self._args.sysctl)
 
-    def init_sysctl(self, description: str) -> None:
-        """Initialize sysctl and listen for connections.
+    def init_sysctl(self, description: str, sysctl_config_file: str) -> None:
+        """Initialize sysctl, listen for connections and configure system.
```
Comment (alexruzenhack) [inline hathor/cli/run_node.py:369]: Suggesting "and apply settings from config file"

File: hathor/builder/sysctl_builder.py
Comment (msbrogli) [inline hathor/builder/sysctl_builder.py:1]: I don't think the file format should be a yaml. I guess a text file with one command per line that allow comments is a simpler solution and also allows one to save the contents of the command `!backup` and use it straight into the `----sysctl-config-file` argument. What do you think?

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/builder/sysctl_builder.py
Comment (alexruzenhack) [reply hathor/builder/sysctl_builder.py:1]: Hm... In the design we use a `txt` extension for this document, however I saw that a `yaml` extension would be easier to implement because we already have a foundation to use this extension, and I didn't want to reinvent the wheel.

Why not make the `!backup` output in a `yaml` style? This way we guarantee this convenience and stay consistent with config documents in yaml style.

By having different configuration documents with different extensions may turn dev experience confusing.

Comment (jansegre) [reply hathor/builder/sysctl_builder.py:1]: I think the concept itself is already a bit confusing, but I understand it is helpful in some cases. To me it would be easier to understand it as `--sysctl-init-file` which is a file that the sysctl subsystem is initialized with, as if it had received the same content as it would through the sysctl socket. It's also better to think of it as something the node is initialized with and not as a configuration.

Comment (alexruzenhack) [reply hathor/builder/sysctl_builder.py:1]: Oh! Ok! I misunderstood when I did the comment in the daily. You are proposing just to change the options name, right? Not to have the sock definition inside it, right?

Comment (msbrogli) [reply hathor/builder/sysctl_builder.py:1]: I agree with @jansegre that `--sysctl-init-file` is more intuitive. Please, update the design as well.

File: hathor/cli/run_node.py [hunk 7]
```diff
@@ -357,10 +359,14 @@ def __init__(self, *, argv=None):
...
-            self.init_sysctl(self._args.sysctl)
+            if self._args.sysctl_config_file:
+                self.init_sysctl(
+                    self._args.sysctl, self._args.sysctl_config_file)
+            else:
+                self.init_sysctl(self._args.sysctl)
 
-    def init_sysctl(self, description: str) -> None:
-        """Initialize sysctl and listen for connections.
+    def init_sysctl(self, description: str, sysctl_config_file: str = None) -> None:
```
Comment (msbrogli) [inline hathor/cli/run_node.py]: `Optional[str]`?

File: hathor/cli/run_node.py [hunk 8]
```diff
@@ -357,10 +359,14 @@ def __init__(self, *, argv=None):
         self.prepare()
         self.register_signal_handlers()
         if self._args.sysctl:
-            self.init_sysctl(self._args.sysctl)
+            if self._args.sysctl_config_file:
```
Comment (msbrogli) [inline hathor/cli/run_node.py:362]: We don't need this if block. Using `self.init_sysctl(self._args.sysctl, self._args.sysctl_config_file)` would work fine since `self._args.sysctl_config_file is None` when a file is not provided.

File: hathor/cli/run_node.py [hunk 9]
```diff
@@ -357,10 +359,14 @@ def __init__(self, *, argv=None):
...
+                self.init_sysctl(
+                    self._args.sysctl, self._args.sysctl_config_file)
+            else:
+                self.init_sysctl(self._args.sysctl)
 
-    def init_sysctl(self, description: str) -> None:
-        """Initialize sysctl and listen for connections.
+    def init_sysctl(self, description: str, sysctl_config_file: str = '') -> None:
```
Comment (msbrogli) [inline hathor/cli/run_node.py:368]: `Optional[str]`?

File: hathor/utils/txt.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import ast
+import os
+from pathlib import Path
+from typing import Any, Union
+
+
+def dict_from_file(*, filepath: Union[Path, str]) -> dict[str, Any]:
```
Comment (msbrogli) [inline hathor/utils/txt.py:21]: Why don't we split the parser from `SysctlProtocol.lineReceived()` to a separate `hathor/sysctl/parser.py` and use it?

File: tests/cli/test_sysctl_init.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+'''
+
+file_invalid_value = '''\
+p2p.rate_limit.global.send_tips=!!tuple [1,2]\
+'''
+
+
+class SysctlInitTest(unittest.TestCase):
+
+    @patch('builtins.open', mock_open(read_data=file_p2p_valid_data))
```
Comment (msbrogli) [inline tests/cli/test_sysctl_init.py:35]: To prevent patching `builtins.open` and `os.path.isfile`, you can simply create a temporary folder, put a file there, test your code, and delete the folder. We do it in many other tests that you can use as a reference.

**Review by msbrogli — CHANGES_REQUESTED** [review]

**Review by jansegre — APPROVED** [review]

File: hathor/builder/sysctl_builder.py [hunk 12]
```diff
@@ -12,14 +12,21 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from typing import Optional
+
 from hathor.builder import BuildArtifacts
+from hathor.conf.settings import HathorSettings
 from hathor.sysctl import ConnectionsManagerSysctl, Sysctl, WebsocketManagerSysctl
 
+settings = HathorSettings
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply hathor/builder/sysctl_builder.py:21]: Removed.

**Review by jansegre — APPROVED** [review]

**Review by jansegre — APPROVED** [review]

File: hathor/builder/sysctl_builder.py [hunk 13]
```diff
@@ -12,14 +12,18 @@
...
 from hathor.builder import BuildArtifacts
 from hathor.sysctl import ConnectionsManagerSysctl, Sysctl, WebsocketManagerSysctl
 
 
 class SysctlBuilder:
     """Builder for the sysctl tree."""
+
     def __init__(self, artifacts: BuildArtifacts) -> None:
         self.artifacts = artifacts
+        self.sysctl_init_file: Optional[str] = None
```
Comment (msbrogli) [inline hathor/builder/sysctl_builder.py:26]: It seems this attribute is not needed anymore.

File: hathor/sysctl/init_file_loader.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class SysctlInitFileLoader:
+    def __init__(self, runner: SysctlRunner, init_file: Optional[str]) -> None:
+        assert runner
+
+        self.runner = runner
+        self.init_file = init_file
+
+    def load(self) -> None:
```
Comment (msbrogli) [inline hathor/sysctl/init_file_loader.py:13]: Missing docstring.

File: hathor/sysctl/init_file_loader.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+        if not self.init_file:
```
Comment (msbrogli) [inline hathor/sysctl/init_file_loader.py:14]: I don't like this behavior of "silently not loading". I feel that this class should always expect an initialization file and fails if none is provided.

File: hathor/sysctl/init_file_loader.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
+from typing import Optional
+
+from hathor.sysctl.runner import SysctlRunner
+
+
+class SysctlInitFileLoader:
+    def __init__(self, runner: SysctlRunner, init_file: Optional[str]) -> None:
```
Comment (msbrogli) [inline hathor/sysctl/init_file_loader.py:7]: I think `init_file` must be a string (instead of an optional string).

File: hathor/sysctl/runner.py [hunk 17]
```diff
@@ -75,6 +82,10 @@ def deserialize(self, value_str: str) -> Any:
         """Deserialize a value sent by the client."""
         if len(value_str) == 0:
             return ()
+
+        if re.match(array_pattern, value_str):
```
Comment (msbrogli) [inline hathor/sysctl/runner.py:86]: This change seems unrelated to this PR. Is it a bug fix? If yes, it's missing in the acceptance criteria. Anyway, I think the fix might be wrong.

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/builder/sysctl_builder.py (hunk 13 above)
Comment (alexruzenhack) [reply hathor/builder/sysctl_builder.py:26]: removed a9e338497c6c5aed78b32589334fd14fe1fde2ff

File: hathor/sysctl/init_file_loader.py (hunk 15 above)
Comment (alexruzenhack) [reply hathor/sysctl/init_file_loader.py:14]: Ok. I made it mandatory: baf31c3c81a310841b7e94c0266c14e2447e2de2

File: hathor/sysctl/init_file_loader.py (hunk 16 above)
Comment (alexruzenhack) [reply hathor/sysctl/init_file_loader.py:7]: Done

File: hathor/sysctl/init_file_loader.py (hunk 14 above)
Comment (alexruzenhack) [reply hathor/sysctl/init_file_loader.py:13]: Done: e61fb0af920a8a22342e9a2cbcf62b067299aa94

File: hathor/sysctl/runner.py (hunk 17 above)
Comment (alexruzenhack) [reply hathor/sysctl/runner.py:86]: Why?

Comment (alexruzenhack) [reply hathor/sysctl/runner.py:86]: But yes, this is a fix for array parsing, and is being tested with [this line](https://github.com/HathorNetwork/hathor-core/pull/714/files#diff-6075f7992b86c5c2974496f5bf63e7f81b8286c675967d8c84c47efff0f690e2R104). Without this we can't deserialize an array with more than 1 element, because it is interpreted as a tuple.

Comment (alexruzenhack) [reply hathor/sysctl/runner.py:86]: I updated the acceptance criteria.

File: hathor/sysctl/init_file_loader.py (hunk 15 above)
Comment (alexruzenhack) [reply hathor/sysctl/init_file_loader.py:14]: run_node adjustment: f5b80e83c66ff705d6069d32a8845c596b7b969e

File: hathor/sysctl/init_file_loader.py (hunk 14 above)
Comment (msbrogli) [reply hathor/sysctl/init_file_loader.py:12]: THanks!

**Review by jansegre — APPROVED** [review]

File: hathor/cli/run_node.py [hunk 18]
```diff
@@ -357,10 +359,11 @@ def __init__(self, *, argv=None):
         self.prepare()
         self.register_signal_handlers()
         if self._args.sysctl:
-            self.init_sysctl(self._args.sysctl)
+            self.init_sysctl(
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/cli/run_node.py]: Thanks!

File: hathor/builder/sysctl_builder.py (hunk 13 above)
Comment (msbrogli) [reply hathor/builder/sysctl_builder.py]: Thanks!

File: hathor/sysctl/init_file_loader.py (hunk 15 above)
Comment (msbrogli) [reply hathor/sysctl/init_file_loader.py]: Thanks!

**Review by msbrogli — APPROVED** [review]

**Review by jansegre — APPROVED** [review]
